### PR TITLE
Remove parasites in any warhead

### PR DESCRIPTION
**Pull request by FS-21**
- Warheads are now able to use this logic for removing parasytes or kick them out into the battlefield.
- If `CanRemoveParasites.KickOut.Paralysis` is negative it will work as `CanRemoveParasites.KickOut.Paralysis=15`.

In `rulesmd.ini`:
```ini
[SOMEWARHEAD]            ; Warhead
CanRemoveParasites=no                   ; boolean
CanRemoveParasites.ReportSound=         ; Sound
CanRemoveParasites.KickOut=no           ; boolean
CanRemoveParasites.KickOut.Paralysis=15 ; integer, game frames
CanRemoveParasites.KickOut.Anim=        ; Animation
```

## Summary by CodeRabbit

- **New Features**
	- Introduced the ability for warheads to remove or kick parasites into the battlefield, enhancing strategic gameplay options.
- **Documentation**
	- Updated documentation to include details on the new parasite removal functionality.
- **Enhancements**
	- Added new properties to warheads for detailed control over parasite removal behavior.

### CODE REVIEW

File: docs/New-or-Enhanced-Logics.md [hunk 1]
```diff
@@ -1180,6 +1180,20 @@ In `rulesmd.ini`:
 NotHuman.DeathSequence=  ; integer (1 to 5)
 ```
 
+### Remove parasites
+- Warheads are now able to use this logic for removing parasites or kick them out into the battlefield.
+- If `CanRemoveParasites.KickOut.Paralysis` is negative it will work as `CanRemoveParasites.KickOut.Paralysis=15`.
```
Comment (MortonPL) [inline docs/New-or-Enhanced-Logics.md:1185]: What's the point of -1? Why not use 15 as the default?

File: docs/Whats-New.md [hunk 2]
```diff
@@ -35,6 +35,7 @@ You can use the migration utility (can be found on [Phobos supplementaries repo]
 - `[JumpjetControls]->TurnToTarget` and `JumpjetTurnToTarget` are obsolete. Jumpjet units who fire `OmniFire=no` weapons **always** turn to targets as other units do.
   - `OmniFire.TurnToTarget` is recommended for jumpjet units' omnifiring weapons for facing turning.
 - Buildings delivered by trigger action 125 will now **always** play buildup anim as long as it exists. `[ParamTypes]->53` is deprecated.
+- Remove parasites logic in any warhead
```
Comment (MortonPL) [inline docs/Whats-New.md:38]: This is in the migration section, not changelog.

File: src/Ext/Techno/Hooks.Shield.cpp [hunk 3]
```diff
@@ -40,6 +41,84 @@ DEFINE_HOOK(0x701900, TechnoClass_ReceiveDamage_Shield, 0x6)
 				RD::SkipLowDamageCheck = true;
 		}
 	}
+
+	if (ScriptExt::IsUnitAvailable(pThis, false))
```
Comment (MortonPL) [inline src/Ext/Techno/Hooks.Shield.cpp:45]: A code block this big should be moved to a function IMO.

Comment (MortonPL) [reply src/Ext/Techno/Hooks.Shield.cpp:45]: There's a couple of nested if statements too, use guard clauses instead.

File: src/Ext/Techno/Hooks.Shield.cpp [hunk 4]
```diff
@@ -40,6 +41,84 @@ DEFINE_HOOK(0x701900, TechnoClass_ReceiveDamage_Shield, 0x6)
...
+	{
+		if (const auto pFoot = abstract_cast<FootClass*>(pThis))
+		{
+			// Ignore other cases that aren't useful for this logic
+			if (pFoot->ParasiteEatingMe
+				&& pThis->WhatAmI() != AbstractType::Infantry
+				&& pThis->WhatAmI() != AbstractType::Building)
```
Comment (MortonPL) [inline src/Ext/Techno/Hooks.Shield.cpp:52]: This looks unnecessary, buildings aren't FootClass and infantry can't be infested (I think?)

File: src/Ext/Techno/Hooks.Shield.cpp [hunk 5]
```diff
@@ -40,6 +41,84 @@ DEFINE_HOOK(0x701900, TechnoClass_ReceiveDamage_Shield, 0x6)
...
+							auto parasiteOwner = parasite->Owner;
+							parasite->IsAlive = false;
+							parasite->IsOnMap = false;
+							parasite->Health = 0;
+
+							parasiteOwner->RegisterLoss(parasite, false);
+							parasiteOwner->RemoveTracking(parasite);
+							parasite->UnInit();
+
+							return 0;
```
Comment (MortonPL) [inline src/Ext/Techno/Hooks.Shield.cpp:96]: This is repeated code, you can use a lambda instead.

File: src/Ext/WarheadType/Body.h [hunk 6]
```diff
@@ -120,6 +120,12 @@ class WarheadTypeExt
 
 		bool PossibleCellSpreadDetonate;
 
+		Nullable<bool> CanRemoveParasites;
```
Comment (MortonPL) [inline src/Ext/WarheadType/Body.h:123]: No need to use Nullable<> if it's a toggle that's false by default.

File: docs/New-or-Enhanced-Logics.md (hunk 1 above)
Comment (FS-21) [reply docs/New-or-Enhanced-Logics.md:1185]: Maybe bad default value decision :-D
Let me change it.

File: src/Ext/Techno/Hooks.Shield.cpp (hunk 4 above)
Comment (FS-21) [reply src/Ext/Techno/Hooks.Shield.cpp]: Infantry is necessary, infantry get parasited and the game makes this insta-kill, I don't know the building case but after discovering what happens with infantry...

File: src/Ext/Techno/Body.h [hunk 7]
```diff
@@ -138,6 +138,7 @@ class TechnoExt
 	static bool ConvertToType(FootClass* pThis, TechnoTypeClass* toType);
 	static bool CanDeployIntoBuilding(UnitClass* pThis, bool noDeploysIntoDefaultValue = false);
 	static bool IsTypeImmune(TechnoClass* pThis, TechnoClass* pSource);
+	static void RemoveParasyte(TechnoClass* pThis, HouseClass* sourceHouse, WarheadTypeClass* wh);
```
Comment (MortonPL) [inline src/Ext/Techno/Body.h:141]:
```suggestion
	static void RemoveParasite(TechnoClass* pThis, HouseClass* sourceHouse, WarheadTypeClass* wh);
```

**Review by MortonPL — CHANGES_REQUESTED** [review]

File: src/Ext/Techno/Body.h (hunk 7 above)
Comment (FS-21) [reply src/Ext/Techno/Body.h]: I though I fixed that in the past... Fixed again!

**Review by MortonPL — APPROVED** [review]

**Review by chaserli — CHANGES_REQUESTED** [review]
* IIRC the game handles parasite suppression in `FootClass::ReceiveDamage`, it would be better to check how that's implemented and rewrite the code there (and perhaps also in `ParasiteClass::AI`) instead of duplicating everywhere. 

* `RegisterLoss` and `RemoveTracking` are not idempotent.

**Comment by FS-21** [comment]
> * IIRC the game handles parasite suppression in `FootClass::ReceiveDamage`, it would be better to check how that's implemented and rewrite the code there (and perhaps also in `ParasiteClass::AI`) instead of duplicating everywhere.
> * `RegisterLoss` and `RemoveTracking` are not idempotent.

I'll convert this PR again as draft until I rework this feature looking those places you mentioned. But for now I'll update it because I need it in my mod.
Thanks for the suggestion!

**Comment by chaserli** [comment]
Already implemented in de2eb053

**Comment by FS-21** [comment]
What happens with the other tags for customizing the feature?

CanRemoveParasites.ReportSound=         ; Sound
CanRemoveParasites.KickOut=no           ; boolean
CanRemoveParasites.KickOut.Paralysis=15 ; integer, game frames
CanRemoveParasites.KickOut.Anim=        ; Animation

**Comment by chaserli** [comment]
> What happens with the other tags for customizing the feature?
> 
> CanRemoveParasites.ReportSound= ; Sound CanRemoveParasites.KickOut=no ; boolean <--- this "no" does what you mentioned in that commit (only remove the parasite). CanRemoveParasites.KickOut.Paralysis=15 ; integer, game frames CanRemoveParasites.KickOut.Anim= ; Animation

Maybe you can try to find a way to detonate a weapon (or AE?) when the parasite gets ejected?

**Comment by FS-21** [comment]
RemoveParasite= only deletes the parasite so I don't understand the ejected suggestion.

Anyways in my mod I only used the sound tag when the parasite is erased by the warhead so I won't make a drama :-D

**Pull request closed without merging**